Public/Src/Engine/Scheduler/Distribution/ExecutionResultSerializer.cs

```diff
@@ -613,5 +621,51 @@ private void WriteFileArtifact(BuildXLWriter writer, FileArtifact file)
             WriteAbsolutePath(writer, file.Path);
             writer.WriteCompact(file.RewriteCount);
         }
+
+        private static void ReadPipProperties(BuildXLReader reader, out IReadOnlyDictionary<string, int> pipProperties)
+        {
+            bool hasPipProperties = reader.ReadBoolean();
+
+            if (!hasPipProperties)
+            {
+                pipProperties = null;
+            }
+            else
+            {
+                int count = reader.ReadInt32Compact();
+
+                var tmpPipProperties = new Dictionary<string, int>();
+
+                for (int i = 0; i < count; i++)
+                {
+                    string key = reader.ReadNullableString();
+                    tmpPipProperties[key] = reader.ReadInt32Compact();
+                }
+
+                pipProperties = tmpPipProperties;
+            }
+        }
+
+        private static void WritePipProperties(BuildXLWriter writer, IReadOnlyDictionary<string, int> pipProperties)
+        {
+            bool hasPipProperties = pipProperties != null && pipProperties.Count != 0;
+
+            if (!hasPipProperties)
+            {
+                writer.Write(false);
+            }
+            else
+            {
+                writer.Write(true);
+
+                writer.WriteCompact(pipProperties.Count);
+
+                foreach (string key in pipProperties.Keys)
+                {
+                    writer.WriteNullableString(key);

```

Comment:
Could the key be `null`?

Reply:
Dictionaries can't have null keys.  The if condition ensures we only hit this block if we have dictionary items.

Reply:
Then, should we simply use `writer.Write(key)` ?

---
In reply to: [334751796](https://github.com/microsoft/BuildXL/pull/1000#discussion_r334751796) [](ancestors = 334751796)